src/Storages/MergeTree/MergeTreePartitionGlobalMinMaxIdxCalculator.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <utility>
+
+#include <Core/Field.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+
+namespace DB
+{
+
+/*
+ * Calculates global min max indexes for a given set of parts.
+ * */
+class MergeTreePartitionGlobalMinMaxIdxCalculator
+{
+    using DataPart = IMergeTreeDataPart;
+    using DataPartPtr = std::shared_ptr<const DataPart>;
+    using DataPartsVector = std::vector<DataPartPtr>;
+public:
+    static std::vector<std::pair<Field, Field>> calculate(
+        const MergeTreeData & storage,
+        const DataPartsVector & parts
+    );
+
+private:
+    static std::vector<std::pair<Field, Field>> extractMinMaxIndexesFromBlock(const Block & block_with_min_max_idx);
+    static std::vector<std::pair<Field, Field>> calculateMinMaxIndexesForPart(const MergeTreeData & storage, const DataPartPtr & part);
+    static void updateGlobalMinMaxIndexes(
+        std::vector<std::pair<Field, Field>> & global_min_max_indexes,
+        const std::vector<std::pair<Field, Field>> & local_min_max_indexes);
```

Comment:
In my personal opinion, in this case, decomposition makes the code harder to read because you need to jump between these functions when reading the code. In particular, I don't see any sense in having `extractMinMaxIndexesFromBlock` as a separate function because it's used only in `calculateMinMaxIndexesForPart` and only as a tail call. Both functions are about 10 lines, it could be just one 20-lines function. But it's okay to keep it this way.

Reply:
Also, it's not Java, you don't have to declare a class if you need just one static function. You can just declare a standalone function in .h file and (if necessary) declare helper functions in a private namespace in the corresponding .cpp file.

This header file exposes unnecessary implementation details increasing build time. If I want to change the signature of some private helper function, I will have to change it in both .h and .cpp files and rebuild all .cpp files that include this header.